Frameworks/ImageIO/CGImageDestination.mm

```diff
@@ -15,11 +16,867 @@
+            RETURN_NULL_IF_FAILED(_idStream->InitializeFromIStream(dataStream));
+        } else {
+            // We are not currently handling data consumer as destination because data consumer is not implemented in WinObjC
+            UNIMPLEMENTED_WITH_MSG("Destination as Data Consumer is not handled right now");
+        }
+
+        RETURN_NULL_IF_FAILED(_idEncoder->Initialize(_idStream.Get(), WICBitmapEncoderNoCache));
+
+        if (_type == typeGIF) {
+            RETURN_NULL_IF_FAILED(_idEncoder->GetMetadataQueryWriter(&_idGifEncoderMetadataQueryWriter));
+        }
+    }
+
+    return self;
+}
+
+@end
+
+// Helper function for setting the property write values when formatting image metadata.
+// Image metadata stores decimal numbers as an integer divided by another integer.
+// The HighPart is a number representing what to divide the LowPart by to get the actual value.
+void setHighLowPartsUnsigned(ULARGE_INTEGER* valueLarge, double valueDouble) {
+    if (valueDouble < 0) {
+        NSTraceInfo(TAG, @"Tried to pass negative value to set unsigned long!\n");
+        return;
+    }
+
+    // Check to see if the value has a decimal component. If not, just divide by 1.
+    if (valueDouble == (unsigned int)valueDouble) {
+        (*valueLarge).LowPart = (unsigned int)valueDouble;
+        (*valueLarge).HighPart = 1;
+    } else {
+        (*valueLarge).LowPart = (unsigned int)(valueDouble * 100);
+        (*valueLarge).HighPart = 100;
+    }
+}
+
+// Signed version of the above
+void setHighLowPartsSigned(LARGE_INTEGER* valueLarge, double valueDouble) {
+    // Check to see if the value has a decimal component. If not, just divide by 1.
+    if (valueDouble == (int)valueDouble) {
+        (*valueLarge).LowPart = (int)valueDouble;
+        (*valueLarge).HighPart = 1;
+    } else {
+        (*valueLarge).LowPart = (int)(valueDouble * 100);
+        (*valueLarge).HighPart = 100;
+    }
+}
+
+// Helper function to write properties to image frames
+void writePropertyToFrame(PROPVARIANT* propertyToWrite, LPCWSTR path, IWICMetadataQueryWriter* propertyWriter) {
+    HRESULT status = propertyWriter->SetMetadataByName(path, propertyToWrite);
+    if (!SUCCEEDED(status)) {
+        NSString* pathNSString = [[NSString alloc] initWithBytes:path
+                                                          length:wcslen(path)*sizeof(wchar_t)
+                                                        encoding:NSUnicodeStringEncoding];
+        NSTraceInfo(TAG, @"Set %@ failed with status=%x\n", pathNSString, status);
+        [pathNSString release];
+    }
+}
+
+void setVariantFromDictionary(CFDictionaryRef dictionary,
+                              CFStringRef key,
+                              VARTYPE propertyType,
+                              LPCWSTR path,
+                              IWICMetadataQueryWriter* propertyWriter) {
+    if (dictionary && CFDictionaryContainsKey(dictionary, key)) {
```

Comment:
Yes, there are places where I call this for a general image property without checking if properties are passed in to AddImage. This performs the role of both checking if there is a properties dictionary as well as checking if the specified key is in it.